Make the url generator available on the events

kbond opened the pull request:
Fixes #286. Tests pass still so hopefully that means this is BC. Before I finalize, I'd like to ensure this direction is acceptable.

TODO:
- [x] Add tests
- [x] Trigger deprecation warnings
- [x] Document

kbond committed:
make the url generator available on the events

kbond commented on src/Event/SitemapPopulateEvent.php:
```diff
@@ -39,13 +40,20 @@ class SitemapPopulateEvent extends Event
+     * @param string|null                $section
+     * @param UrlGeneratorInterface|null $urlGenerator
+     */
+    public function __construct(UrlContainerInterface $urlContainer, string $section = null, UrlGeneratorInterface $urlGenerator = null)
```
Thinking we could trigger a deprecation warning if `$urlGenerator` is null - WDYT?

yann-eugone replied:
Yes, 100% agreed

kbond commented on src/Event/SitemapAddUrlEvent.php:
```diff
@@ -49,13 +50,20 @@ class SitemapAddUrlEvent extends Event
+     * @param array<string, mixed>       $options
+     * @param UrlGeneratorInterface|null $urlGenerator
+     */
+    public function __construct(string $route, array $options, UrlGeneratorInterface $urlGenerator = null)
```
I assumed you would want this even to include the `$urlGenerator` as well. If so, we should trigger a deprecation warning here when it is null.

yann-eugone replied:
Not sure what would be the use case here, but why not.

But yes I think a deprecation should be triggered here if `$urlGenerator` is null.

yann-eugone replied:
Night passed, I'm wondering whether these deprecations at the event level are making sense.

We agree that this is a DX thing ?
A developer that rely on an event the bundle produces will be able to use this feature.
But if he do not want to use this feature and produces is own event, we don't care, do we ?

kbond replied:
Sure, that's fair. I'll remove.

kbond commented on src/Service/AbstractGenerator.php:
```diff
@@ -45,20 +46,27 @@ abstract class AbstractGenerator implements UrlContainerInterface
+     * @param UrlGeneratorInterface|null $urlGenerator
      */
-    public function __construct(EventDispatcherInterface $dispatcher, int $itemsBySet = null)
+    public function __construct(EventDispatcherInterface $dispatcher, int $itemsBySet = null, UrlGeneratorInterface $urlGenerator = null)
```
Trigger deprecation when `null` here?

yann-eugone replied:
Yes, here too

yann-eugone commented on src/Event/SitemapAddUrlEvent.php:
```diff
@@ -123,4 +131,9 @@ public function getOptions(): array
         return $this->options;
     }
+
+    public function getUrlGenerator(): ?UrlGeneratorInterface
```
IMO we should have a strict return type, and throw an exception if property is null.

Something that tell the developer what he did wrong.

kbond replied:
Perfect, I thought the same thing.

yann-eugone commented on src/Event/SitemapPopulateEvent.php:
```diff
@@ -65,4 +73,9 @@ public function getSection(): ?string
         return $this->section;
     }
+
+    public function getUrlGenerator(): ?UrlGeneratorInterface
```
Same has above: strict return type, and throw an exception if property is null.

yann-eugone reviewed (changes requested):
Sorry I'm late (long weekend :))

I think this is a very good start, thank you.

Do you planned to update/add tests in the same PR ?

kbond commented:
> Do you planned to update/add tests in the same PR ?

Yes, that was my plan.

kbond commented:
Ok, addressed the comments. Let me know if you'd like me to change the deprecation/exception messages.

For the tests, I was thinking the following:
1. Mock/inject the `UrlGeneratorInterface` in the unit tests to remove the self-deprecations
2. Adjust the integration test's `SitemapListener` to use the url generator from the event (instead of injecting)

Let me know if you'd like anything else.

yann-eugone commented on composer.json:
```diff
@@ -17,7 +17,8 @@
         "symfony/console": "^4.4|^5.0",
-        "symfony/framework-bundle": "^4.4|^5.0"
+        "symfony/framework-bundle": "^4.4|^5.0",
+        "symfony/deprecation-contracts": "^2.4"
```
I'm not a big fan of adding a dependency in a minor version.
We usually have issues with dependencies version compatibility : X requires Y 1.x, but Z requires Y 2.x

So far we did not used this library, we were triggering deprecations manually.
Do you think you can go back to that strategy ?

kbond replied:
Sure, I can do this.

yann-eugone commented:
>     1. Mock/inject the `UrlGeneratorInterface` in the unit tests to remove the self-deprecations

I know that this bundle has some bad practices in tests, but I'm trying to fix this.
I'm trying to apply the "don't mock what you don't own" approach.
I suggest you inject an implementation of the UrlGeneratorInterface, some actual object from symfony ecosystem, or a test dummy you will create.

>     2. Adjust the integration test's `SitemapListener` to use the url generator from the event (instead of injecting)

:+1:, 100% agree with this

kbond committed:
adjust integration test to use url generator from event

kbond committed:
fix self deprecation

kbond committed:
added tests

kbond committed:
adjust docs

kbond commented:
Ok, I believe this is ready for a final review. I believe the phpstan errors are unrelated.

kbond committed:
fix cs

yann-eugone reviewed (approved):
Thank you for this work !
